### PR TITLE
chore: trim flat tools/list catalog + remove tools/list pagination (#181)

### PR DESCRIPTION
**Pull request by kingpanther13**
## Summary

- **Pagination removed from `tools/list`.** PR #190 paginated it at page size 50, but many MCP clients (including Claude.ai's connector) do NOT iterate `nextCursor` and silently saw only the first 50 tools. Replaced with a single-response handler; backstopped by the existing universal response-size guard at `handleMcpRequest()` (124,000-byte threshold) that emits a loud `-32603 "Response too large"` envelope if the catalog ever exceeds the hub cap. **Loud failure beats silent client truncation.**
- **Flat catalog trimmed back under the cap.** The catalog was already over-cap territory on `main` (~125-131 KB) — pagination was masking that. Removing pagination only works if the catalog fits, so wrapped enumeration / reference blocks in four heavy tools in `[[FLAT_TRIM]] ... [[/FLAT_TRIM]]` markers, lifted to the existing TOOL_GUIDE.md sections.
- **Conditional transform, not a hard prose lift.** Gateway-mode callers (the default) see the schema unchanged. The transform strips content between markers ONLY in the flat-mode `tools/list` path; every other client-visible surface (gateway-catalog mode, `search_tools` corpus, missing-param error hints) strips just the marker tokens so wrapped content stays available.
- Cursor pagination on `tools/call` (`list_devices`, `list_installed_apps`, `list_rm_rules`, etc. via `_paginateList`) is **unchanged** — that path is genuinely opt-in and remains.

## Type of change

- [ ] `feat` — new feature or capability
- [ ] `fix` — bug fix
- [x] `chore` — maintenance, dependency bump, or housekeeping
- [ ] `refactor` — code restructure with no behaviour change
- [ ] `docs` — documentation only
- [ ] `test` — tests only
- [ ] `ci` — CI/CD pipeline change

(Marking `chore` since the bulk is the trim work; the pagination removal is technically a `fix` for a regression in #190, called out in the Summary.)

## Changes

- `hubitat-mcp-server.groovy`:
  - **`handleToolsList`**: removed the cursor/`nextCursor` pagination block. Returns the full catalog in a single response. Stale clients that pass a `cursor` get the full catalog and find no `nextCursor` (graceful migration). Inline comment documents the removal rationale.
  - New helpers `stripFlatTrim(String, boolean)` + `applyDescriptionTransform(List, boolean)` co-located with `getToolDefinitions()`. Two-pass regex handles both own-line marker blocks (drop with bracketing newlines) and inline markers (drop content/keep token boundaries). Markers must be balanced and non-nested. In flat-mode (`dropContent=true`) the regexes require both markers, so a stray open or close survives unchanged and the CI marker-leakage guard trips loud (no silent data loss). In token-strip mode (`dropContent=false` — gateway catalog, search corpus, missing-param hint) a lone marker is silently removed, since the marker token is itself the bug to prevent on those surfaces.
  - Wired the transform into all four client-visible surfaces: `getToolDefinitions()` flat branch (drop content), `getToolDefinitions()` gateway branch (strip tokens only, defensive), `handleGateway()` catalog mode + pre-validate missing-param hint (strip tokens), `buildToolSearchCorpus()` (strip tokens — defensive against a future *top-level* tool description ever gaining markers; today's corpus only indexes top-level descriptions plus gateway summary text, not the property descriptions where this PR's markers actually live).
  - Wrapped enumerations in four tools:
    - `update_native_app`: `addTrigger.description` capability families, `addAction.description` capability families, `addRequiredExpression.description` STPage capability list (inline mid-bullet).
    - `list_devices`: response-shape paragraph + server-side-filter paragraph in the top-level description; valid-name enumeration inside the `fields` parameter description.
    - `get_hpm_drift`: drift-signal-type definitions, response-field enumeration, `signals[]` entry shape, data-quality warning types, partial-detection suffix, under-count caveats — all duplicating existing TOOL_GUIDE.md content.
    - `create_native_app`: `appType` bullet list + multi-paragraph PARTIAL-SUCCESS HANDLING block.
  - Replaced each wrapped section with a tightened pointer to `{discover: true}` and/or the relevant TOOL_GUIDE.md subsection.
- `TOOL_GUIDE.md`: new `#### update_native_app capability reference` and `#### create_native_app reference` subsections under `### manage_native_rules_and_apps (12 tools)`. The `list_devices` and `get_hpm_drift` schema pointers reuse existing TOOL_GUIDE entries (Performance Tips and `manage_hpm` respectively). The `### tools/list Pagination` section rewritten to describe the new single-response contract.
- `SKILL.md`: updated the `tools/list` bullet describing the no-pagination contract; opt-in `tools/call` cursor story unchanged.
- `tests/BAT-v2.md`: T606 rewritten as "tools/list returns the full flat catalog in one response (no pagination)" — exercises the new contract, asserts no `nextCursor`, asserts every flat-mode tool present, asserts stale cursors silently ignored.
- New `src/test/groovy/server/UpdateNativeAppSchemaTrimSpec.groovy`: 11 Spock features covering the flat catalog byte budget (with diagnostic per-tool breakdown on failure), marker leakage in both modes, byte delta between flat vs gateway disclosure surfaces, anchor preservation in flat mode, gateway-catalog content preservation, `search_tools` BM25 corpus contract (defensive guard for top-level descriptions), missing-param hint contract, `applyDescriptionTransform` idempotence, schema↔TOOL_GUIDE anchor integrity, and direct unit tests for the `stripFlatTrim` helper including empty-content and unmatched-marker edge cases.
- Updated `HandleMcpRequestDispatchSpec`: the 6 cursor-based tests (iterating-cursors flat-mode, gateway single-page, 3 cursor-rejection cases, empty-cursor) collapsed into 3 tests pinning the new contract — single-response in both modes + cursor-is-ignored-regardless-of-value (data-driven across numeric, non-numeric, out-of-range, negative, empty).

Part of #181. (The issue scope was a broader audit of heavy tools; this PR trims the top 4 by JSON size — `update_native_app`, `get_hpm_drift`, `create_native_app`, `list_devices` — and removes the tools/list pagination footgun that was masking the catalog growth.)

## Release Notes

- Internal: removed pagination from `tools/list` (was added in #190 — caused silent catalog truncation on clients that didn't iterate `nextCursor`); now returns the full catalog in a single response. Backstopped by the existing `-32603 "Response too large"` envelope.
- Internal: trimmed the flat-mode (`useGateways=false`) `tools/list` descriptions for `update_native_app`, `get_hpm_drift`, `create_native_app`, and `list_devices` to keep the flat catalog under the hub's 124 KB JSON-RPC response cap. No user-visible change in the default gateway mode.

## Testing

Per-property and per-description byte deltas after anchor tightening + the 3-tool extension:

| Tool / param | Gateway | Flat | Saved |
|---|---:|---:|---:|
| `update_native_app.addTrigger` | ~6,400 B | ~3,450 B | ~2,950 B |
| `update_native_app.addAction` | ~11,200 B | ~4,400 B | ~6,800 B |
| `update_native_app.addRequiredExpression` | ~4,250 B | ~3,600 B | ~650 B |
| `list_devices` top-level | ~1,320 B | ~520 B | ~800 B |
| `list_devices.fields` | (in inputSchema) | (trim inline) | ~700 B |
| `get_hpm_drift` top-level | ~4,130 B | ~1,160 B | ~2,960 B |
| `create_native_app` top-level | ~3,070 B | ~1,660 B | ~1,400 B |
| **Total** | | | **~16 KB** |

Catalog impact: flat catalog ~115 KB post-trim, ~9 KB cap headroom — comfortable margin for the pagination removal to be safe.

Regression guards in the new spec:
- Flat catalog ≤ 121,000 bytes (the ~3 KB cap-headroom goal from #181).
- Failure path reports the 5 heaviest tools by JSON size so a future regression points at where to start trimming.
- No `[[FLAT_TRIM]]` token leaks into either flat-mode or gateway-mode `tools/list` JSON.
- `update_native_app`'s flat-mode entry is at least 8 KB smaller than its gateway-catalog entry.
- Flat-mode descriptions preserve the `{discover: true}` / TOOL_GUIDE.md anchors.
- Gateway-catalog response from `handleGateway('manage_native_rules_and_apps', null, [:])` keeps the full capability enumerations.
- `search_tools` BM25 corpus contains no marker tokens — defensive against a future *top-level* tool description gaining markers.
- Missing-param error hint for `update_native_app` strips marker tokens but preserves the wrapped capability prose.
- `applyDescriptionTransform` is idempotent — a second call on the same in-place mutated list is a no-op.
- Schema-named anchor headers exist in `TOOL_GUIDE.md` (catches a future rename of a TOOL_GUIDE subsection silently breaking the in-schema pointer).
- `stripFlatTrim` unit tests cover own-line/inline markers, dropContent true/false, null input, no-marker passthrough, empty marker content, and the unmatched-marker asymmetry.

Regression guards on the pagination removal (in `HandleMcpRequestDispatchSpec`):
- `tools/list` with `useGateways=false` returns the full flat catalog in one response with NO `nextCursor`, every expected tool name present, no duplicates.
- `tools/list` with `useGateways=true` returns the full gateway catalog in one response with NO `nextCursor`.
- `tools/list` ignores any `cursor` parameter a stale client passes (data-driven across numeric, non-numeric, out-of-range, negative, empty values) — full catalog returned, no error, no `nextCursor`.

## Checklist

- [x] **Unit tests added for any new MCP tools** (required — see [docs/testing.md](docs/testing.md) for the harness + recipes)
- [x] Sandbox lint passes: `python tests/sandbox_lint.py`
- [ ] `./gradlew test` passes locally (or CI confirms)
- [ ] Live-hub BAT tests updated if tool behaviour changed (see `tests/BAT-v2.md` — T606 was updated)
- [x] Documentation updated if user-facing behaviour or tool surface changed

### CODE REVIEW

File: TOOL_GUIDE.md [hunk 1]
```diff
@@ -556,6 +556,79 @@ If a user asks "create a new RM rule" or "modify this Room Lighting instance":
...
+  - `ifThen` + `expression={...}` (opens IF block; close with `endIf`)
+  - `elseIf` + `expression={...}` (continues IF block; needs preceding `ifThen`)
+
+##### `addRequiredExpression` STPage capability list
+
+RM 5.1 Required Expression conditions accept these capability values (per-condition `capability` field):
+
+`Switch`, `Motion`, `Contact`, `Lock`, `Presence`, `Smoke detector`, `Water sensor`, `Tamper alert`, `Acceleration`, `Carbon monoxide detector`, `Carbon dioxide sensor`, `Power source`, `Mode`, `Private Boolean`, `Custom Attribute`, `Battery`, `Dimmer`, `Energy meter`, `Fan Speed`, `Humidity`, `Illuminance`, `Power meter`, `Temperature`, `Thermostat cool setpoint`, `Thermostat fan mode`, `Thermostat heat setpoint`, `Thermostat mode`, `Thermostat state`, `Window Shade`, `Days of week`, `Between two dates`, `Between two times`, `On a Day`, `Last Event Device`, `Lock codes`.
```
[the remark this answers is not in the document]
Comment (kingpanther13) [reply TOOL_GUIDE.md]: Done in b6e49f6 — reformatted the `addRequiredExpression` STPage capability list into grouped bullets (Device-state / Numeric / Time-based / Hub state / Custom or other) matching the `addTrigger` and `addAction` section style.

**Review by kingpanther13 — COMMENTED** [review]
Addressed Gemini review feedback in b6e49f6 — `addRequiredExpression` STPage capability list in TOOL_GUIDE.md now uses grouped bullets matching the `addTrigger` / `addAction` formats.
